arrow/obsidianmodule.c

```diff
@@ -214,9 +214,16 @@ Obsidian_evolve(ObsidianObject *self, PyObject *args)
   substrates[0] = self->substrates_count;
 
   // Create new python numpy arrays from the raw C results
-  PyObject *time_obj = PyArray_SimpleNewFromData(1, steps, NPY_DOUBLE, result.time);
-  PyObject *events_obj = PyArray_SimpleNewFromData(1, steps, NPY_INT64, result.events);
-  PyObject *outcome_obj = PyArray_SimpleNewFromData(1, substrates, NPY_INT64, result.outcome);
+  PyArrayObject *time_obj = (PyArrayObject *) PyArray_SimpleNew(1, steps, NPY_DOUBLE);
+  memcpy(PyArray_DATA(time_obj), result.time, (sizeof (double)) * result.steps);
+  PyArrayObject *events_obj = (PyArrayObject *) PyArray_SimpleNew(1, steps, NPY_INT64);
+  memcpy(PyArray_DATA(events_obj), result.events, (sizeof (long)) * result.steps);
+  PyArrayObject *outcome_obj = (PyArrayObject *) PyArray_SimpleNew(1, substrates, NPY_INT64);
+  memcpy(PyArray_DATA(outcome_obj), result.outcome, (sizeof (long)) * self->substrates_count);
```

Comment:
Great work debugging this! FYI Python 3 improves memory debugging features, e.g. `--with-valgrind` compile option and `faulthandler` module.

In retrospect, we should've thought of this issue since numpy can't know what deallocator to call or whether you pointed it to a portion of a larger memory node.

(FYI, it looks like one [can register a destructor](https://stackoverflow.com/a/52737023/1682419), but letting numpy handle the memory is simpler.)

Please make the code check for memory allocation failures.

Technically `result.events` (et al) should be declared as an array of `int64_t`, not `long` whose size varies with the C compiler, the target CPU architecture, and compiler flags.

It's better to use `sizeof(result.events[0])` (et al) so it can't get out of sync with the declaration.

Similarly, `steps` and `substrates` should be arrays of `npy_intp`, although in this case the compiler should complain if the element size is wrong.

Reply:
> In retrospect, we should've thought of this issue since numpy can't know what deallocator to call or whether you pointed it to a portion of a larger memory node.

Yeah you mentioned this exact section as well. I was going off of the big red warning box in the numpy C api docs here: https://docs.scipy.org/doc/numpy-1.15.0/reference/c-api.array.html, which were correct, but my assumption was not. Good find on the destructor! But yes, I agree this is simpler. 

> Technically `result.events` (et al) should be declared as an array of `int64_t`, not `long` whose size varies with the C compiler, the target CPU architecture, and compiler flags.

Okay, I'll give that a shot.